lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/FileBasedSystemAccessControl.java

```diff
@@ -811,20 +824,32 @@ public void checkCanSetMaterializedViewProperties(SystemSecurityContext context,
     @Override
     public void checkCanGrantExecuteFunctionPrivilege(SystemSecurityContext context, String functionName, TrinoPrincipal grantee, boolean grantOption)
     {
+        Identity identity = context.getIdentity();
+        boolean allowed = functionRules.stream()
+                .filter(functionRule -> functionRule.matches(identity.getUser(), identity.getEnabledRoles(), identity.getGroups(), functionName))
+                .map(CatalogFunctionAccessControlRule::canGrantExecuteFunction)
+                .findFirst()
+                .orElse(false);
+        if (!allowed) {
+            String granteeAsString = format("%s '%s'", grantee.getType().name().toLowerCase(ENGLISH), grantee.getName());
+            denyGrantExecuteFunctionPrivilege(functionName, context.getIdentity(), granteeAsString);
+        }
     }
 
     @Override
     public void checkCanGrantExecuteFunctionPrivilege(SystemSecurityContext context, FunctionKind functionKind, CatalogSchemaRoutineName functionName, TrinoPrincipal grantee, boolean grantOption)
     {
-        switch (functionKind) {
-            case SCALAR, AGGREGATE, WINDOW:
-                return;
-            case TABLE:
-                // TODO (https://github.com/trinodb/trino/issues/12833) implement
-                String granteeAsString = format("%s '%s'", grantee.getType().name().toLowerCase(ENGLISH), grantee.getName());
-                denyGrantExecuteFunctionPrivilege(functionName.toString(), context.getIdentity(), granteeAsString);
+        AccessMode requiredCatalogAccess = functionKind == TABLE ? ALL : READ_ONLY;
```

Comment:
I think it should be more something like this:

```suggestion
        AccessMode requiredCatalogAccess = switch(functionKind) {
            case SCALAR, AGGREGATE, WINDOW -> READ_ONLY;
            case TABLE -> ALL;
        };
```

Reply:
What if we add new function kind in a future?

Reply:
That's exactly the point. This is an exhaustive switch, so if there is a new enum value, this will stop compiling and you'll be forced to consider what to do. This the same future-proofing that @findepi did in the same place earlier.

Reply:
Got it, applied.